File: 124-binary-tree-maximum-path-sum/124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    public int maxPath(TreeNode node, int maxValue[]) {
        if (node == null) return 0;
        int left = Math.max(0, maxPath(node.left, maxValue));
        int right = Math.max(0, maxPath(node.right, maxValue));
        maxValue[0] = Math.max(maxValue[0], left + right + node.val);
        return Math.max(left, right) + node.val;
    }
    
    public int maxPathSum(TreeNode root) {
        int maxValue[] = new int[1];
        maxValue[0] = Integer.MIN_VALUE;
        maxPath(root, maxValue);
        return maxValue[0];
    }
}

File: 124-binary-tree-maximum-path-sum/124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    int maxPath(TreeNode* root, int &maxSum) {
        
        if(root == NULL) return 0;
        
        // max is used to return only +ve values from the noes ie -ve nodes will be stored as 0 
        int leftSum = max(0, maxPath(root->left, maxSum));
        int rightSum = max(0, maxPath(root->right, maxSum));
        
        maxSum = max(maxSum, (root->val + leftSum + rightSum));
        
        return root->val + max(leftSum, rightSum);
    }
    
    int maxPathSum(TreeNode* root) {
        
        int maxSum = INT_MIN;
        maxPath(root, maxSum);
        return maxSum;
    }
};